### PR TITLE
Add fold_docs for DocId list

### PR DESCRIPTION
**Pull request by garrensmith**
## Overview
Adds a fold_docs function that will fetch the docs for a list of
Doc Id's in parallel. This is used for _all_docs?keys=["id1", "id2"].

## Testing recommendations

All existing tests should pass

## Checklist

- [x] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/chttpd/src/chttpd_db.erl [hunk 1]
```diff
@@ -902,9 +902,10 @@ send_all_docs_keys(Db, #mrargs{} = Args, VAcc0) ->
         _ -> Args#mrargs.doc_options
     end,
     IncludeDocs = Args#mrargs.include_docs,
-    lists:foldl(fun(DocId, Acc) ->
-        OpenOpts = [deleted | DocOpts],
-        Row0 = case fabric2_db:open_doc(Db, DocId, OpenOpts) of
+    OpenOpts = [deleted | DocOpts],
+
+    CB = fun({doc, DocId, Doc}, Acc) ->
```
Comment (davisp) [inline src/chttpd/src/chttpd_db.erl:907]: This `{doc, DocId, Doc}` seems a bit awkward. Maybe better as `fun(DocId, Doc, Acc) -> ... end`?

File: src/fabric/src/fabric2_db.erl [hunk 2]
```diff
@@ -969,6 +970,145 @@ fold_docs(Db, UserFun, UserAcc0, Options) ->
...
+fold_docs(Db, DocIds, UserFun, UserAcc0, Options) ->
+    fabric2_fdb:transactional(Db, fun(TxDb) ->
+        try
+            NeedsTreeOpts = [revs_info, conflicts, deleted_conflicts],
+            NeedsTree = (Options -- NeedsTreeOpts /= Options),
+
+            FetchRevs = case NeedsTree of
+                true ->
+                    fun(DocId) ->
+                        fabric2_fdb:get_all_revs_future(TxDb, DocId)
```
Comment (davisp) [inline src/fabric/src/fabric2_db.erl:982]: I don't think you've included an implementation of `get_all_revs_future` in fabric2_fdb unless I'm really blind. I'd probably add it as part of the refactoring commit.

File: src/fabric/src/fabric2_db.erl [hunk 3]
```diff
@@ -969,6 +970,145 @@ fold_docs(Db, UserFun, UserAcc0, Options) ->
...
+            } = FinalAcc3,
+            {ok, FinalUserAcc}
+
+        catch throw:{stop, StopUserAcc} ->
+            {ok, StopUserAcc}
+        end
+    end).
+
+
+drain_fold_docs_revs_futures(_TxDb, #{revs_count := C} = Acc) when C < 100 ->
```
Comment (davisp) [inline src/fabric/src/fabric2_db.erl:1026]: All of these support functions should be moved to the end of the module. I'd put them just after `drain_all_deleted_info_futures`.

File: test/elixir/test/all_docs_test.exs [hunk 4]
```diff
@@ -242,13 +262,17 @@ defmodule AllDocsTest do
     resp = Couch.post(
       "/#{db_name}/_all_docs",
       body: %{
-        :keys => [1, 2],
-        :limit => 1
+        :keys => ["1", "2"],
```
Comment (davisp) [inline test/elixir/test/all_docs_test.exs:265]: Why did we have to change the keys to strings here? Is there an API change here that we should make a note of?

Comment (davisp) [reply test/elixir/test/all_docs_test.exs:265]: I notice the previous test is still using integer keys so I assume those still work. The addition of testing include_docs is nice, but I'd prefer to not change unrelated bits if possible.

**Comment by davisp** [comment]
Also, it looks like we're only using `get_winning_revs_wait` in four spots. We should probably change those to `get_revs_wait` instead of keeping the `get_winning_revs_wait` alias function. That'll help keep cruft to a minimum rather than letting it accumulate slowly.

File: src/chttpd/src/chttpd_db.erl (hunk 1 above)
Comment (garrensmith) [reply src/chttpd/src/chttpd_db.erl]: Fixed

File: src/fabric/src/fabric2_db.erl (hunk 2 above)
Comment (garrensmith) [reply src/fabric/src/fabric2_db.erl:982]: Good spot. Fixed

File: src/fabric/src/fabric2_db.erl (hunk 3 above)
Comment (garrensmith) [reply src/fabric/src/fabric2_db.erl]: Fixed

File: test/elixir/test/all_docs_test.exs (hunk 4 above)
Comment (garrensmith) [reply test/elixir/test/all_docs_test.exs:265]: The original test was just returning `%{"error" => "not_found", "key" => 1}` which is not quite what we should be testing. We should be testing we return the document for that key. So I've added in the test to check that we return an error for the incorrect key and a proper test to prove we return keys. No functionality changed. Just improving the tests.

Comment (davisp) [reply test/elixir/test/all_docs_test.exs:265]: Ah! Sounds good. That certainly wasn't testing what I thought it was testing by reading it.

**Pull request merged**